File: backend/dataset/coinChange/coinChange_18.py
class Solution:
    def coinChange(self, coins:List[int], amount:int)->bool:
        his, coins, queue, step = set(), set(coins), {0}, 1
        
        if amount == 0: return 0
        if amount in coins: return 1
        
        while queue:
            new_queue = set()
            for now in queue:
                for i in coins:
                    this = now+i
                    if this in his or this>amount: 
                        continue
                    else:
                        his.add(this)

                    if amount-this in coins:
                        return step+1
                    else:
                        new_queue.add(this)
                        
            queue = new_queue
            step += 1
        return -1

File: backend/CDataset/coinChange/coinChange_18.cpp
int coinChange(vector<int>& coins, int amount) { 
        int dp[amount+1];
        dp[0] = 0;    
        for(int e=1;e<amount+1;e++)
            dp[e] = amount+1;
        for(int i=1;i<amount+1;i++){
            for(int j:coins){
                if(i-j>=0)
                    dp[i] = min(dp[i],dp[i-j]+1);
            }
        }
        if(dp[amount]==amount+1)
            return -1;
        
        return dp[amount];
    }
